### ISSUE TITLE
Karabiner virtual driver activation does not work on macOS 12.0.1

### ISSUE THREAD

**Issue by yyoshiho** [pqrs-org/Karabiner-Elements#2876]
After Karabiner upgraded to 14.2.0 from 13.7.0 on MacBook Pro 13-inch 2017 running macOS 12.0.1, Karabiner-Elements Alert dialog was displayed to deactivate the driver. It seemed deactivation was successful, but this dialog is still appeared for deactivation after macOS was rebooted, not for activation (no activate button).
Then I had to downgrade to 13.7.0. This version is still working with Monterey 12.0.1 without any problem.

**Comment by tekezo** [comment]
Thank you for feedback!

It seems that Karabiner driver upgrade did not work for some reason.
If you are familiar with the command line, could you show me the result of the following command on Terminal after Karabiner-Elements upgrade?

**Show DriverKit status command**

```shell
systemextensionsctl list
```

**Upgrade DriverKit Driver command**

```shell
/Applications/.Karabiner-VirtualHIDDevice-Manager.app/Contents/MacOS/Karabiner-VirtualHIDDevice-Manager activate
```

**Comment by yyoshiho** [comment]
Requested results after upgraded to 14.2.0:

% systemextensionsctl list 
2 extension(s)
--- com.apple.system_extension.driver_extension
enabled	active	teamID	bundleID (version)	name	[state]
*	*	G43BCU2T37	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice (1.0.0/1.0.0)	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice	[activated enabled]
--- com.apple.system_extension.endpoint_security
enabled	active	teamID	bundleID (version)	name	[state]
*	*	X9E956P446	com.crowdstrike.falcon.Agent (6.30/143.01)	Agent	[activated enabled]
% 

and

% sudo /Applications/.Karabiner-VirtualHIDDevice-Manager.app/Contents/MacOS/Karabiner-VirtualHIDDevice-Manager activate
activation of org.pqrs.Karabiner-DriverKit-VirtualHIDDevice is requested
org.pqrs.Karabiner-DriverKit-VirtualHIDDevice will be replaced to 1.5.0 from 1.0.0
request of org.pqrs.Karabiner-DriverKit-VirtualHIDDevice is failed with error: The operation couldn’t be completed. (OSSystemExtensionErrorDomain error 10.)
%

**Comment by tekezo** [comment]
Thank you for information!

`OSSystemExtensionErrorDomain error 10` is `forbiddenBySystemPolicy `.
So, loading new driver is blocked by some system policy.

Does CrowdStrike Falcon provide a feature to be able to block system extension?
If so, I guess it causes this issue. Please confirm the CrowdStrike Falcon configuration.

**Comment by yyoshiho** [comment]
Actually I'm not sure because this mac is managed by jamf. So I can not look CrowdStrike settins. But version 13.7.0 can be reinstalled without any manual operations.
Original 13.7.0 version had been upgraded before this mac was upgraded to 12.0.1 Monterey. I'm not sure it is related or not. Anyway I will continue to use 13.7.0 for some time to come.
Thanks for your help.

**Comment by mirosval** [comment]
I have a similar problem:

```
❯ systemextensionsctl list
1 extension(s)
--- com.apple.system_extension.driver_extension
enabled	active	teamID	bundleID (version)	name	[state]
		G43BCU2T37	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice (1.5.0/1.5.0)	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice	[activated waiting to upgrade]
```

and I'm unable to uninstall using `systemextensionsctl unistall G43BCU2T37 org.pqrs.Karabiner-DriverKit-VirtualHIDDevice` or via the UI. I also dont see the activation button, so I'm stuck. I have restored an older intel based mac backup onto a new machine with apple silicon.

**Comment by tekezo** [comment]
@mirosval 

Uninstalling DriverKit is a very strict operation required by macOS.
The version of the loaded cached driver must match the version on the file system.

1. You have to install Karabiner-Elements 14.2.0 to install v1.5.0 driver.
2. Unload the driver by uninstaller.
   https://karabiner-elements.pqrs.org/docs/manual/operation/uninstall/
3. Install any versions of Karabiner-Elements.

**Comment by terwin** [comment]
I have the same issue as @yyoshiho after upgrading Karabiner-Elements to version 14.2 — repeated alerts to deactivate the old driver and reboot. I saw the issue on  macOS 11.4, and it persists after upgrading to macOS 11.6.1.  Also, like @yyoshiho, my Mac is managed by Jamf.

**Comment by yyoshiho** [comment]
I tried to isolate the problem and I uninstalled the version 13.7.0, then 13.7.0 also couldn't activate.
  activate error was "OSSystemExtensionErrorDomain error 10" and
  deactivate error was "OSSystemExtensionErrorDomain error 4".
I also tried,
- Resetting your Mac's PRAM and NVRAM
  https://support.apple.com/kb/ht1379
- Starting up in Safe Mode (Safe Mode purges cache files)
  https://support.apple.com/kb/ht1455
and
1. Restart in recovery mode (keep pressing Cmd+R).
2. Open terminal, type spctl kext-consent add G43BCU2T37.
But both workarounds for Big-sur doesn't help for Monterely.

Some news mentioned some drivers based on DriverKit Extension:.dext doesn't work on Monterey 12.0.1, I don't know it affects to Karabiner virtual keyboard  driver. But I hope Karabiner will work with new coming Monterey release in the future.

**Comment by HuskyNie** [comment]
I have same issue as @yyoshiho after upgrad 14.2
```shell
systemextensionsctl list
2 extension(s)
--- com.apple.system_extension.driver_extension
enabled	active	teamID	bundleID (version)	name	[state]
*	*	G43BCU2T37	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice (1.0.0/1.0.0)	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice	[activated enabled]
--- com.apple.system_extension.network_extension
enabled	active	teamID	bundleID (version)	name	[state]
*	*	-	com.adguard.mac.adguard.network-extension (2.5.3/955)	AdGuard Network Extension	[activated enabled]
```
```shell
/Applications/.Karabiner-VirtualHIDDevice-Manager.app/Contents/MacOS/Karabiner-VirtualHIDDevice-Manager activate

activation of org.pqrs.Karabiner-DriverKit-VirtualHIDDevice is requested
org.pqrs.Karabiner-DriverKit-VirtualHIDDevice will be replaced to 1.5.0 from 1.0.0
request of org.pqrs.Karabiner-DriverKit-VirtualHIDDevice is failed with error: The operation couldn’t be completed. (OSSystemExtensionErrorDomain error 8.)
```

**Comment by inamuu** [comment]
I have same issue. but I solved this could not disable driver by workaround.

- Karabiner-Elements to 14.2
- macOS 12.0.1 on Intel chip

```sh
$ systemextensionsctl uninstall G43BCU2T37 org.pqrs.Karabiner-DriverKit-VirtualHIDDevice
At this time, this tool cannot be used if System Integrity Protection is enabled.
This limitation will be removed in the near future.
Please remember to re-enable System Integrity Protection!
```

So, I tried to below

1. Uninstall karabiner-Elements(but I couldn't uninstall driver ver 1.0.0)
2. disable SIP
3. Install karaviner-Elements(driver version 1.5.0) -> OK
4. enable SIP


https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection

**Comment by flybayer** [comment]
I managed to get it working with this

1. Installed 14.2.2 - found it was broken
2. Download and install 13.7
3. 13.7 wasn't working, complaining about driver mismatch
4. Uninstalled 13.7 from the button inside the app
5. Installed 13.7 - not working, same driver mismatch
6. Upgrade to 14.2.2 from inside the app
7. Now everything is working on 14.2.2

**Comment by givihuda** [comment]
I'm running into the issue with macOS Big Sur 11.x after following the update prompt from Karabiner-Elements (upon Mac startup) to upgrade the app; I now have version 14.2.0 (which "Check for Updates" says is the latest). My iHome Bluetooth keyboard is back to its initial broken state (due to it being an "international" keyboard sold in the US).

I get the pop-up saying my virtual keyboard driver is **_outdated_** (not disabled) and will be upgraded if I follow the instructions; step 1 says to deactivate it, I then click the button to deactivate it. Step 2 says to Restart the Mac. I do. I get the pop-up again. Step 3 says to activate the virtual keyboard driver, but there's no instructions on how to do this. I don't know where to do this or I'd try. I'm afraid to run commands at the moment because I don't understand if they'll work the way I expect on my Big Sur. I'm not going to upgrade to Monterey for a while because upgrading to Big Sur from Catalina broke some apps I rely on for some months. I'm afraid to upgrade anything on Mac at this point, Apple keeps changing things so much and locking down things so much.

![image](https://user-images.githubusercontent.com/74785256/141814779-5820f817-706d-42a2-8653-f3a5e7e2d4bf.png)

I'm going to try @flybayer instructions... though I don't know where he got 14.2.2 from, since it appears to be newer than the latest.

However, I don't understand his step 4: Uninstall from button inside the app. I see no button inside the app to uninstall.

**Comment by givihuda** [comment]
Uninstalling the app from the Misc tab Uninstall button, then re-installing my old downloaded version (13) fixed the issue. I was in an endless cycle deactivating the virtual keyboard driver from the KE pop-up, where Step 3 gave no instructions on how to activate the driver.

**Comment by tsuyo** [comment]
For whom having the original "CrowdStrike Falcon" issue (whose `$ systemextensionsctl list` shows com.crowdstrike.falcon.Agent enabled) like me, most likely the issue would be resolved by removing this. In summary,

1. Uninstall Karabiner
2. Uninstall Falcon (Your IT might set a security token to uninstall this. In that case, ask them for it)
```
$ sudo /Applications/Falcon.app/Contents/Resources/falconctl uninstall
```
3. Reboot Mac
4. Just in case, check `$ systemextensionsctl list` shows nothing
5. Install Karabiner

**Comment by givihuda** [comment]
I don't have CrowdStrike, but I do have Avast free:

```~ % systemextensionsctl list
3 extension(s)
--- com.apple.system_extension.driver_extension
enabled	active	teamID	bundleID (version)	name	[state]
*	*	G43BCU2T37	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice (1.0.0/1.0.0)	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice	[activated enabled]
--- com.apple.system_extension.network_extension
enabled	active	teamID	bundleID (version)	name	[state]
*	*	6H4HRTU5E3	com.avast.Antivirus.SystemExtension (3.1.99/3.1.99)	Avast Antivirus System Extension	[activated enabled]
*	*	DE8Y96K9QP	com.cisco.anyconnect.macos.acsockext (4.10.03104/4.10.03104)	Cisco AnyConnect Socket Filter Extension	[activated enabled]```

I don't know if this is helpful or not.

I'm commenting on this issue because it's closer to the issue I'm having, whereas #2596 is from Jan 25 and I seem to not even get that far yet... I mean, are these duplicate issues, or slightly different? I don't understand everything I'm reading there.

**Comment by Allon-Guralnek** [comment]
I'm getting `OSSystemExtensionErrorDomain error 8` as well. Haven't managed to get Karabiner working since upgrading to v14. Can't seem to be able to deactivate the driver.

```
4 extension(s)
--- com.apple.system_extension.network_extension
enabled	active	teamID	bundleID (version)	name	[state]
*	*	TZ3UEPFYKD	com.checkpoint.fw.filter (871010711/6)	Check Point Firewall	[activated enabled]
*	*	UBF8T346G9	com.microsoft.wdav.netext (101.47.27/101.47.27)	Microsoft Defender ATP Network Extension	[activated enabled]
--- com.apple.system_extension.endpoint_security
enabled	active	teamID	bundleID (version)	name	[state]
*	*	UBF8T346G9	com.microsoft.wdav.epsext (101.47.27/101.47.27)	Microsoft Defender ATP Endpoint Security Extension	[activated enabled]
--- com.apple.system_extension.driver_extension
enabled	active	teamID	bundleID (version)	name	[state]
*	*	G43BCU2T37	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice (1.0.0/1.0.0)	org.pqrs.Karabiner-DriverKit-VirtualHIDDevice	[activated enabled]
```

**Comment by raine** [comment]
> I'm getting OSSystemExtensionErrorDomain error 8 as well. Haven't managed to get Karabiner working since upgrading to v14. Can't seem to be able to deactivate the driver. I'm on Monterey 12.0.1 (21A559) on an Intel MacBook.

I believe I had this issue and the solution was to boot to recovery mode, deactivate SIP, reboot to normal mode, deactivate the driver, activate the driver and then once more boot to recovery mode to activate SIP again.

**Comment by givihuda** [comment]
I downgraded my Karabiner-Elements while ago, and it's been working fine on version 13.4.0 on macOS Big Sur 11.6.2. I'm afraid to upgrade, though I'm replacing my keyboard and mouse soon.

**Comment by RiZKiT** [comment]
I'm afraid there is no other option, because disabling SIP in Recovery Mode is not always an option, if you have a managed MacBook with unknown root password. I'm currently happy after downgrading to 13.7.0, but will be unable to update.

**Comment by CampAsAChamp** [comment]
> For whom having the original "CrowdStrike Falcon" issue (whose `$ systemextensionsctl list` shows com.crowdstrike.falcon.Agent enabled) like me, most likely the issue would be resolved by removing this. In summary,
> 
> 1. Uninstall Karabiner
> 2. Uninstall Falcon (Your IT might set a security token to uninstall this. In that case, ask them for it)
> 
> ```
> $ sudo /Applications/Falcon.app/Contents/Resources/falconctl uninstall
> ```
> 
>  3. Reboot Mac
> 4. Just in case, check `$ systemextensionsctl list` shows nothing
> 5. Install Karabiner

@tsuyo 
Did you reinstall Falcon? I'm a little hesitant to try this because I don't want to just uninstall Falcon (I'm guessing its an IT monitoring tool) and potentially screw up whatever was set up on my computer before I got it. I'm using a company computer.

**Comment by tsuyo** [comment]
@CampAsAChamp I didn't even notice this fact, but Falcon is currently enabled and active again (and I don't have any issue right now). So this should be one time operation.

**Comment by Slumber86** [comment]
I have the same problem, but my it doesn't want to let me uninstall falcon. Do you have some other workarounds.

**Comment by givihuda** [comment]
I got rid of the keyboard that relied on Elements, because the keyboard had its own issues, so I've uninstalled Elements.

I'm just gonna stick with an Apple long keyboard, as overpriced as I think they are, because they just offer the best compatibility with Mac (except for not being able to assign volume buttons or much else to the F16-19 keys, but that's a different issue... or maybe Elements could enable that?!).

Anyway, I really dislike having to install 3rd-party software to do what Apple should be allowing to begin with, so I consider the coder of this app to be more intelligent than Apple in regards to keyboard configuration.

**Comment by yyoshiho** [comment]
Today, I upgraded my MacBook Pro (13" 2017) to 12.4. This reported issue has been resolved.
At first, I had thought Karabiber-Elements driver is restricted by Jamf policy. But Jamf did not restrict the driver.

**Comment by bgmort** [comment]
For those who are finding this issue, if your system is managed by JAMF, #2636 may have the solution to your problem.

**Comment by nejmeijer** [comment]
It seems the original issue has been resolved. Feel free to open a new issue if what is already mentioned here does not cover your case.
